File: solution/473/answer.cpp
class Solution {
public:
    bool makesquare(vector<int>& nums) {
        if(nums.size()<4)return false;
        int sum = 0;
        for(int i=0;i<nums.size();i++)sum+=nums[i];
        if(sum%4!=0)return false;
        int target = sum/4;
        int allnum = 1 << nums.size();
        vector<int> oresult;
        for(int x=0;x<allnum;x++){
            int itemsum = 0;
            for(int i=0;i<nums.size();i++){
                if(1<<i&x){itemsum+=nums[i];}
            }
            if(itemsum==target)oresult.push_back(x);
        }
        vector<int> tresult;
        for(int x=0;x<oresult.size();x++){
            for(int i=0;i<oresult.size();i++){
                if(oresult[x]&oresult[i])continue;
                tresult.push_back(oresult[x]|oresult[i]);
            }
        }
        for(int x=0;x<tresult.size();x++){
            for(int i=0;i<tresult.size();i++){
                if(tresult[x]&tresult[i])continue;
                return true;
            }
        }
        return false;
    }
};

File: solution/473/answer.py
class Solution(object):
    def makesquare(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if len(nums) < 4:return False
        nums.sort(reverse=True)
        num_s = sum(nums)
        if num_s % 4 != 0:return False
        target = num_s / 4
        if nums[0] < target and nums[0] + nums[-1] > target:return False
        oresult = []
        sresult = []
        allnum = 1 << len(nums)
        for i in xrange(allnum):
            isum = 0
            for x in xrange(len(nums)):
                if 1<<x & i:
                    if nums[x] > target:return False
                    isum += nums[x]
                    if isum > target:break
            if isum == target:oresult.append(i)
        for x in oresult:
            for y in oresult:
                if x & y:continue
                sresult.append(x | y)
        for x in sresult:
            for y in sresult:
                if x & y:continue
                return True
        return False

s = Solution()
print s.makesquare([4,8,12,16,20,24,28,32,36,40,44,48,52,56,60])